Activate Devise::Lockable

citizen428 opened the pull request:
## What type of PR is this? (check all applicable)

- [X] Feature

## Description

It was recently brought to our attention that we're not locking out users who attempt to conduct dictionary and brute force attacks against user accounts. This PR activates `Devise::Lockable` to address this.

## QA Instructions, Screenshots, Recordings

1. Make sure your user has a password that you know:
    ```ruby
    user.update(password: "...", password_confirmation: "...")
   ```
2. Go to `/users/sign_in` or `/enter` in your local environment.
3. Enter the wrong password a couple of times and see your account get locked out.
4. You can unlock it in the console with `user.unlock_access!`.

## Added tests?

- [X] yes

## Added to documentation?

- [X] no documentation needed

citizen428 commented on config/initializers/devise.rb:
```diff
@@ -216,14 +216,14 @@
   # :both  = Enables both strategies
   # :none  = No unlock strategy. You should handle unlocking by yourself.
-  # config.unlock_strategy = :both
+  config.unlock_strategy = :both
```
This together with `config.unlock_in = 1.hour` means the account will automatically unlock after 1 hour. We can change that if people feel like it, but since our main motivation for this was deterring brute-force attacks it seemed like an ok setting.

citizen428 commented on config/initializers/devise.rb:
```diff
@@ -216,14 +216,14 @@
   # Number of authentication tries before locking an account if lock_strategy
   # is failed attempts.
-  # config.maximum_attempts = 20
+  config.maximum_attempts = 3
```
20 attempts seemed high. Let me know if you think 3 is too low.

Ridhwana replied:
Personal but not strong opinion: I know that it's common for banks to use 3 but I feel like we can be a bit more lenient and maybe use 5.

citizen428 replied:
Sure, I had 5 at first, then lowered it for local testing, happy to increase it again. Would like to see what others think.

mstruve replied:
I like 5, 3 seems too low to me as well.

citizen428 commented on spec/system/user_logs_in_with_password_spec.rb:
```diff
@@ -0,0 +1,49 @@
+require "rails_helper"
+
+RSpec.describe "Authenticating with a password" do
```
I think we had no specs for password logins so far, so I added a whole bunch, not only the ones for account locking.

Ridhwana replied:
Nice!🎉

Ridhwana commented on spec/system/user_logs_in_with_password_spec.rb:
```diff
@@ -0,0 +1,49 @@
+      allow(User).to receive(:maximum_attempts).and_return(1)
+
+      submit_login_form(user.email, "wr0ng")
+      expect(page).to have_text("Your account is locked.")
```
This is not necessarily related to your PR, but I was curious to know can we configure this to specify how long your account will be locked for. i.e. "You account is locked. Please try again in an hour". This will reduce the amount of support requests and panic we get from users.

citizen428 replied:
It's the option `config.unlock_in` in `config/initializers/devise.rb`. Note that time-based unlocking is actually a convenience feature, normally the account would just be locked until you click the unlock link in the email.

Ridhwana commented on spec/system/user_logs_in_with_password_spec.rb:
```diff
@@ -0,0 +1,49 @@
+      submit_login_form(user.email, password)
+      expect(page).to have_current_path("/dashboard?signin=true")
+    end
+  end
```
Question: Is it possible and useful to test that the text "Your account is locked" is removed in an hour (once the account is unlocked) by doing some test time magic, or do you think that's not necessary to test because it's encompassed by devise?

citizen428 replied:
The text only appears in a flash message after the failed login attempt, not permanently on the page.

rhymes commented:
Hi @justinlarson, can you please take a look at this? Thank you!

justinlarson commented:
Using the 1 hour lockout is ok if you allow valid users to unlock the account through email or logging in with github/twitter. If you don't allow valid users to unlock their account then an attacker _could_ essentially perform an account Denial of Service and lockout that user for at least that hour.

citizen428 commented:
@justinlarson That's how `config.unlock_strategy = :both` for Devise works. The user will receive an email with an unlock token (similar to how password reset works in Devise), but the lock will also automatically expire.

citizen428 commented:
@mstruve Sorry if this was unclear, but that's not how `:both` is supposed to work since `:lockable` has no knowledge about `omniauthable`. As per the initializer, the two supported ways for unlocking are:

1. Via clicking a link in the email
2. Time based

Good call on an adding this to `/internal`.

mstruve commented:
> Via clicking a link in the email

How do you get that link?

citizen428 commented:
> How do you get that link?

Devise should email it when the account gets locked. I'll verify that and potentially add a spec for it.

> One more question, could a person get locked out if they didn't have a password set?

Technically I think all our accounts have passwords, as we set them to `Devise.friendly_token(20)` during social login. I would expect users who can't remember ever setting a password to use the forgot password functionality, but worst case, yes, they can get locked out.

Anyway, I had already set this PR back to draft and am working on some related things now, will re-request a review when I'm done.

citizen428 commented:
Made the following changes now:

1. Increased max attempts to 5 before locking out the user.
2. Verified that an email gets sent and added a spec for it.
    <img width="978" alt="Screen Shot 2020-07-29 at 10 32 17" src="https://user-images.githubusercontent.com/47985/88754605-51200880-d189-11ea-888f-bbea1e2cb097.png">
3. Updated our `Authenticator` class so social logins now unlock locked accounts automatically. Tried locally and added a spec for it.
4. Show the lock status on the internal user show page.
    <img width="280" alt="Screen_Shot_2020-07-29_at_10_33_14" src="https://user-images.githubusercontent.com/47985/88754684-7f9de380-d189-11ea-8beb-85d8618333c5.png">
5. Added a button for unlocking an account and added a spec for it.
    <img width="426" alt="Screen Shot 2020-07-29 at 10 45 23" src="https://user-images.githubusercontent.com/47985/88754720-9f350c00-d189-11ea-9269-fc3c99ac0f93.png">
    <img width="1155" alt="Screen Shot 2020-07-29 at 10 45 31" src="https://user-images.githubusercontent.com/47985/88754735-a65c1a00-d189-11ea-96fd-80e7974a9194.png">

I'll re-request reviews from everyone.

citizen428 commented on db/schema.rb:
```diff
@@ -472,7 +472,6 @@
     t.datetime "updated_at", null: false
     t.bigint "user_id"
     t.datetime "verified_at"
-    t.index ["user_id"], name: "index_email_authorizations_on_user_id"
```
This keeps popping up for me locally, does anyone know anything about it?

mstruve replied:
I had this issue and updated my database manually via a console to get it to go away. I always figure it came from a branch I tested that never ended up in master

citizen428 replied:
Yes, that fixed it. I just tried to avoid a similar situation to what we had a few months back were different PRs alternatively added and then dropped a specific index. 😆

mstruve commented on app/services/authentication/authenticator.rb:
```diff
@@ -127,6 +127,7 @@ def default_user_fields
 
     def update_user(user)
       user.tap do |model|
+        user.unlock_access! if user.access_locked?
```
😍

citizen428 commented:
Cleaned up the schema now.
